src/sagemaker/predictor_async.py

```diff
@@ -0,0 +1,372 @@
+        if self.sagemaker_session.s3_client is None:
+            self.s3_client = self.sagemaker_session.boto_session.client(
+                "s3",
+                region_name=self.sagemaker_session.boto_region_name,
+            )
+        else:
+            self.s3_client = self.sagemaker_session.s3_client
+
+        self.serializer = predictor.serializer
+        self.deserializer = predictor.deserializer
+        self.name = name
+        self._endpoint_config_name = None
+        self._model_names = None
+        self._context = None
+        self._input_path = None
+
+    def predict(
+        self,
+        data=None,
+        input_path=None,
+        initial_args=None,
+        inference_id=None,
+        waiter_config=WaiterConfig(),
+    ):
+        """Wait and return the Async Inference result from the specified endpoint.
+
+        Args:
+            data (object): Input data for which you want the model to provide
+                inference. If a serializer was specified in the encapsulated
+                Predictor object, the result of the serializer is sent as input
+                data. Otherwise the data must be sequence of bytes, and the
+                predict method then sends the bytes in the request body as is.
+            input_path (str): Amazon S3 URI contains input data for which you want
+                the model to provide async inference.
+            initial_args (dict[str,str]): Optional. Default arguments for boto3
+                ``invoke_endpoint_async`` call. (Default: None).
+            inference_id (str): If you provide a value, it is added to the captured data
+                when you enable data capture on the endpoint (Default: None).
+            waiter_config (sagemaker.async_inference.waiter_config.WaiterConfig): Configuration
+                for the waiter. The pre-defined value for the delay between poll is 15 seconds
+                and the default max attempts is 60
+        Raises:
+            ValueError: If both input data and input Amazon S3 path are not provided
+        Returns:
+            object: Inference for the given input. If a deserializer was specified when creating
+                the Predictor, the result of the deserializer is
+                returned. Otherwise the response returns the sequence of bytes
+                as is.
+        """
+        if data is None and input_path is None:
+            raise ValueError(
+                "Please provide input data or input Amazon S3 location to use async prediction"
+            )
+        if data is not None:
+            input_path = self._upload_data_to_s3(data, input_path)
+
+        self._input_path = input_path
+        response = self._submit_async_request(input_path, initial_args, inference_id)
+        output_location = response["OutputLocation"]
+        result = self._wait_for_output(output_path=output_location, waiter_config=waiter_config)
+
+        return result
+
+    def predict_async(
```

Comment:
As discussed last time, we think seperate them will be less painful for customers. As they can keep their code to use `predict` method and they can `predict_async` to avoid blocking on the other side. Also, dividing them will avoid different returning types in one function. This may be better in some use cases.